### ISSUE TITLE
ByteString.utf8CharSequence()

### ISSUE THREAD

**Issue by swankjesse** [square/okio#156]
Would be interesting to return a `CharSequence` that's backed by the same `ByteArray`, but doesn't do UTF-8 to UTF-16 transcoding.

**Comment by coreyalexbarker** [comment]
Is this still something that's needed or wanted?

**Comment by swankjesse** [comment]
The original idea was to build something that doesn’t actually allocate a string, on the theory that doing so could be more efficient. Instead it’d UTF-8 decode on the fly. I don’t think this is feasible though, because we don’t have a mechanism to track the byte offset for a given char offset.

**Comment by coreyalexbarker** [comment]
Ooh, I see. Was the idea to avoid allocating just strings, or any extraneous objects?

**Comment by swankjesse** [comment]
Avoiding allocating strings, avoiding converting from UTF-8 to UTF-16.

**Comment by coreyalexbarker** [comment]
Here's a possible solution. Ignore the messed up imports, they are fixed. This function returns the same char[] as the String constructor would, without creating any strings or leaving UTF-8. It could probably be optimized, but just wanted to get your opinion.

**Comment by swankjesse** [comment]
Neat! The runtime CPU + memory cost of that will be very similar to what we already do. 

I was hoping we could do character transcoding lazily, but this issue is iceboxed cause there's no great API for that. We'd want a cursor for the byte sequence but we don't have a way to expose that.

**Comment by coreyalexbarker** [comment]
That would be extremely useful. I'll think about that and get back to you.

**Comment by swankjesse** [comment]
Cool. Please don’t rush out and write a lot of code; I’m not sure we have a use case.

**Comment by coreyalexbarker** [comment]
It's actually something I'm working on anyway, building CRUD statements from legacy data that's in all kinds of different formats. I found your project while looking for an existing parsing library. I haven't found anything that fits perfectly, so I went ahead and made a ByteCursor class that does what I needed, but not so specific that it couldn't be used elsewhere. You're welcome to check it out.

It accepts a byte array and keeps it's own internal index(user modifiable). It can break the array into separate tokens based on a delimiter that can be changed on each iteration. A few other things too. I'll be cleaning it up, but basic functionality is there.
